src/globus_sdk/experimental/tokenstorage/base.py

```diff
@@ -148,3 +180,47 @@ def user_only_umask(self) -> t.Iterator[None]:
             yield
         finally:
             os.umask(old_umask)
+
+
+def _default_globus_app_filename(
+    client_id: UUIDLike, app_name: str, environment: str
+) -> str:
+    r"""
+    Construct a default TokenStorage filename for a globus app.
+    The filename will have no file format suffix.
+
+    On Windows, this will be:
+        ``~\AppData\Local\globus\app\{client_id}\{app_name}\tokens``
+
+    On Linux and macOS, we use:
+        ``~/.globus/app/{client_id}/{app_name}/tokens``
+    """
+    environment_prefix = f"{environment}-"
+    if environment == "production":
+        environment_prefix = ""
+    filename = f"{environment_prefix}tokens"
+    app_name = _slugify_app_name(app_name)
+
+    if sys.platform == "win32":
+        # try to get the app data dir, preferring the local appdata
+        datadir = os.getenv("LOCALAPPDATA", os.getenv("APPDATA"))
+        if not datadir:
+            home = os.path.expanduser("~")
+            datadir = os.path.join(home, "AppData", "Local")
+        return os.path.join(
+            datadir, "globus", "app", str(client_id), app_name, filename
+        )
+    else:
+        return os.path.expanduser(
+            f"~/.globus/app/{str(client_id)}/{app_name}/{filename}"
+        )
+
+
+def _slugify_app_name(app_name: str) -> str:
+    """
+    Slugify a globus app name for use in a file path.
+
+    This "slugification" will replace spaces with hyphens and lowercase the string so:
+      "My App" -> "my-app"
+    """
+    return app_name.replace(" ", "-").lower()
```

Comment:
`"My  App"` will become `"my--app"`, which is maybe okay. Does Auth have restrictions on the characters used?

For example, is there anything preventing question marks or angle brackets before this potentially reaches the filesystem?

Reply:
Auth does not have any obvious restrictions for us to replicate locally ([thread w/ Josh](https://globus.slack.com/archives/C03N0A8J7/p1722377481158659)).

I don't take issue with `"My  App"` -> `"my-app"`.
We're doing some very modest modification to the app name to present directories a little bit cleaner but we shouldn't be doing much at all here.

Reply:
I'd like to strengthen this. Could I recommend something like:

```python
import re
import unicodedata

app_name = "I <3 Globus!"

path_name = app_name
path_name = "".join(
    c if unicodedata.category(c).startswith(("L", "N")) else "-" for c in path_name
)
path_name = re.sub("-+", "-", path_name).strip("-") or f"app-name-{len(app_name)}"
path_name = path_name.lower()

# path_name == "i-3-globus"
```

(I've broken the above into separate conceptual transformation steps.)

I don't want to have users accidentally creating directories with colons or brackets, and restricting the Unicode character class to letters and numbers seems permissive enough. The other transforms (replacing runs of hyphens, and lowercasing the whole string) aren't strictly important to me, but disallowing forbidden filesystem characters matters a lot to me.

Reply:
Updated the slugify method to meet the following rules:

* Reserved filesystem characters are replaced with a '+'. ('a?' -> 'a+')
* Periods and Spaces are replaced with a '-'. ('a. b' -> 'a--b')
* Control characters are removed. ('a\0b' -> 'ab')
* The string is lowercased. ('AB' -> 'ab')

Additionally an error is raised if the resulting slug is 
* The empty string
* A reserved windows filename ('CON', 'NUL', 'LPT1', etc.)